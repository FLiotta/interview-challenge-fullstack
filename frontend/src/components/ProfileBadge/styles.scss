@import '../../styles/colors.scss';

.profilebadge {
  display: flex;
  align-items: center;
  border-radius: 6px;
  padding: 8px 15px;
  cursor: pointer;
  transition: 0.5s all;
  position: relative;

  &:hover {
    background-color: #eee;
  }

  &--active {
    background-color: #eee;
  }

  &__username {
    font-family: 'montserrat';
    margin-bottom: 0px;
    margin-right: 15px;
  }

  &__menu {
    position: absolute;
    width: 100%;
    background-color: #fff;
    border: 1px solid $brand_color;
    top: 55px;
    left: 0px;
    border-radius: 6px;
    overflow: hidden;

    &-item {
      width: 100%;
      padding: 15px;
      color: $brand_color;
      margin-bottom: 0px;
      display: flex;
      align-items: center;

      &:hover {
        background-color: #eee;
      }

      &-text {
        margin-left: 10px;
        margin-bottom: 0;
      }
    }
  }
}