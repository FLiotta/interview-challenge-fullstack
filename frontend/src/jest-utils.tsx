// @Packages
import { ReactNode } from 'react';
import { render as rtlRender } from '@testing-library/react';
import { Provider } from 'react-redux';

// @Project
import store from 'store';
import { JsxElement } from 'typescript';

function render(
  ui: ReactNode | JsxElement | any,
  renderOptions: any = { }
) {
  function Wrapper({ children }: any) {
    return (
      <Provider store={store}>{children}</Provider>
    )
  }
  return rtlRender(ui, {
     wrapper: Wrapper, 
     ...renderOptions
    })
}

export * from '@testing-library/react'

export { render }
