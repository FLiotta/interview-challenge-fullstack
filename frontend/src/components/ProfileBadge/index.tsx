// @Packages
import React, { ReactElement, ReactNode, useRef, useState } from 'react';
import { IoLogOutOutline } from 'react-icons/io5';
import { BsGear } from 'react-icons/bs';
import { useSelector, useDispatch } from 'react-redux';
import { JsxElement } from 'typescript';
import { useHistory } from 'react-router-dom';
import cn from 'classnames';

// @Project
import { useClickOutside } from 'hooks/useClickOutside';
import { disconnect } from 'actions/session';
import { selectProfile } from 'selectors/profile';
import Avatar from 'components/Avatar';

// @Own
import './styles.scss';
import { IThunkDispatch } from 'interfaces';

export interface Option {
  title: string
  action: () => void
  icon?: ReactElement | JsxElement | ReactNode,
  onlyStaff?: boolean
}

interface IMenuProps {
  options: Option[]
}

export const Menu: React.FC<IMenuProps> = ({ options }) => {
  const profile = useSelector(selectProfile);

  return (
    <div
      className="profilebadge__menu"
      data-testid="profilebadge-menu-testid"
    >
      {options
        .filter((opt) => {
          if(opt.onlyStaff && !profile.is_staff) {
            return false;
          }

          return true;
        })
        .map((opt) => (
          <div
            className="profilebadge__menu-item" 
            onClick={opt.action}
            key={`menuitem__${opt.title}`}
          >
            {opt.icon}
            <p className="profilebadge__menu-item-text">{opt.title}</p>
          </div>
        ))
      }
    </div>
  )
}

const ProfileBadge: React.FC<any> = () => {
  const [menuVisible, setMenuVisible] = useState<boolean>(false);
  const profileBadgeRef = useRef<any>();
  const profile = useSelector(selectProfile);
  const history = useHistory();
  const dispatch = useDispatch();

  useClickOutside(profileBadgeRef, () => { 
    setMenuVisible(false)
  });

  const options: Option[] = [
    {
      title: 'Backoffice',
      action: () => {
        history.push('/panel')
      },
      icon: <BsGear />,
      onlyStaff: true
    },
    {
      title: 'Cerrar sesiÃ³n',
      action: () => {
        dispatch(disconnect())
        history.push('/auth')
      },
      icon: <IoLogOutOutline />
    },
  ]

  return (
    <div
      data-testid="profilebadge-testid"
      className={cn("profilebadge", { "profilebadge--active": menuVisible })}
      onClick={() => setMenuVisible(!menuVisible)}
      ref={profileBadgeRef}
    >
      {menuVisible && <Menu options={options} />}
      <p className="profilebadge__username">{profile.first_name} {profile.last_name}</p>
      <Avatar userId={profile.id} height={30} width={30} />
    </div>
  )
}

export default ProfileBadge;