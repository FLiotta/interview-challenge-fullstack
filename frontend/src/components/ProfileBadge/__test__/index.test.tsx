// @Packages
import { getByTestId } from '@testing-library/react';
import { render, screen, done, fireEvent } from 'jest-utils';

// @Project
import ProfileBadge, { Menu, Option } from '../index';

describe('ProfileBadge component', () => {
  it('Should match snapshot', () => {
    const { container } = render(<ProfileBadge />);

    expect(container).toMatchSnapshot();
  });

  it('Should open menu on click', () => {
    const profileBadgeTestId = 'profilebadge-testid';
    const profileBadgeMenuTestId = 'profilebadge-menu-testid';

    const { container } = render(<ProfileBadge />);

    const profileBadge = getByTestId(container, profileBadgeTestId);

    fireEvent.click(profileBadge);

    const profileBadgeMenu = getByTestId(container, profileBadgeMenuTestId);
    
    expect(profileBadgeMenu).not.toBeNull();
  });

  it('Menu receives "options" prop', () => {
    const OPTIONS: Option[] = [
      {
        title: "Opcion 1",
        action: () => { }
      }
    ];

    const { container } = render(<Menu options={OPTIONS} />);
    const optionsNodes = container.getElementsByClassName("profilebadge__menu-item");

    expect(optionsNodes.length).toBe(OPTIONS.length)
  });

  it('Menu option action must be fired on click', () => {
    const MOCK_FN = jest.fn();
    const OPTIONS: Option[] = [
      {
        title: "Opcion 1",
        action: MOCK_FN
      }
    ];

    const { container } = render(<Menu options={OPTIONS} />);
    const optionNode = container.getElementsByClassName("profilebadge__menu-item")[0];

    fireEvent.click(optionNode);

    expect(MOCK_FN).toHaveBeenCalledTimes(1);
  })
})