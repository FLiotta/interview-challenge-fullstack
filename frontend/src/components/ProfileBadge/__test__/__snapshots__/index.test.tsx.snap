// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileBadge component Should match snapshot 1`] = `
<div>
  <div
    class="profilebadge"
    data-testid="profilebadge-testid"
  >
    <p
      class="profilebadge__username"
    >
       
    </p>
    <div
      class="avatar"
      data-testid="avatar-testid"
      style="width: 30px; height: 30px;"
    >
      <div
        class="avatar__fill avatar__fill--1"
        data-testid="avatar-fill-testid"
      />
    </div>
  </div>
</div>
`;
